ramips: Asus RT-AX53U partition layout change

kvuorine opened the pull request:
Change Asus RT-AX53U partition layout to maximize available space.

Increase DEVICE_COMPAT_VERSION and add DEVICE_COMPAT_MESSAGE to prevent unintentional damage.

kvuorine commented:
I ran into a problem with this after sysupgrading factory image flashed from bootloader which ran fine.

```
RT-AX53U bootloader version: 1.0.0.1
Reading from 0x1e0000, size 0x100000
Succeeded
MAC Address: XX:XX:XX:XX:XX:XX
Reading from 0x1e0000, size 0x40000
Succeeded
Check Factory's EEPROM successfully! (r = 17)
Reading from 0x2e0000, size 0x40000
Succeeded
Check Factory2's EEPROM successfully! (r = 17)
Reading from 0x3e0000, size 0x2000
Succeeded
Error: size '0x7375739f' is too large
nmbm nmbm0 read 0x83000000 0x3e0000 0x7375739f, error !!
Reading from 0x35e0000, size 0x2000
Succeeded
Reading from 0x35e0000, size 0x41
Succeeded
Bad TRX magic.

## Both firmware damaged ...
```

Have to take a closer look.

Djfe commented:
Thanks for taking care of this :)

Djfe commented:
Not a clue, yet.
What where the differences between factory and sysupgrade? Did both have the same dts?

lukasz1992 commented:
Same dts.

sysupgrade is for updating from openwrt, factory for flashing either from openwrt or stock/oem firmware.
Usually sysupgrade has: empty blocks for jffs2, fake-squashfs etc, to allow oem bootloader to boot openwrt.

btw. I am wondering why we have all three partitions: kernel, ubi and also jffs2?
Good example of concating jffs2 and firmware(1) is here: https://github.com/openwrt/openwrt/blob/main/target/linux/ramips/dts/mt7621_dlink_dap-x1860-a1.dts
we could use jffs2 space for ubi

kvuorine commented:
@Djfe do you know if jffs2 partition is used for something, if I remember correctly you added it at some point?

I possibly found my user error which caused sysupgrade to fail, build still going on.

Djfe commented:
Great, I missed that but yes you need to replicate all changes the original Asus commit did. I entirely forgot about the file that is relevant for sysupgrades.

> do you know if jffs2 partition is used for something, if I remember correctly you added it at some point?

No I don't know it, but my device worked fine without it and openwrt definitely doesn't need it.
It could be that Asus even resets the partition on factory reset, but I don't know that. But it should still run fine, if you return to stock rom (I haven't tested that)
I know that when I returned to stock rom before including jffs, my stock settings where all still present. the changed Passwort as well. My guess is that jffs2 is pretty much the partition Asus stores config data in (password, is wifi turned on, dhcp or static ip, etc.).

ok.you just figured this out yourself ^^

kvuorine commented:
This has now been tested by flashing openwrt-factory.bin over oem firmware with mtd-write method and after testing that, sysupgrade -n openwrt-sysupgrade.bin. Seems to work fine.

csharper2005 commented:
Are you sure it's good practice to create different variations for the same device? `ubootmod, stock, ubootmod-alt1, stock-alt2, stock-alt3-ubi`, etc. 

This may:
- Confuse users (so we need clearly described info in the commit message about installing, recovery, which variants are compatible between each other, etc)
- Require additional testing
- Consume infrastructure capacity for the build and storage

Therefore, I would suggest to use this approach for the justified exceptions only. Why not just choose the optimal option?

kvuorine commented:
> Therefore, I would suggest to use this approach for the justified exceptions only. Why not just choose the optimal option?

@csharper2005 I don't like this approach either, original issue which caused me to do this this was with unsolved upgrade path problems from earlier versions of OpenWRT causing bootloops when sysupgrading to a new partition layout. Other developers opted to keep partition table compatible with earlier versions to avoid need for recovery and planned to create another firmware with all flash usable with OpenWRT.

I had to do something to get more usable flash space, if nothing changes in other device firmware I'll have to use this for my devices. Just put it here to see if anyone else sees it beneficial.

csharper2005 commented:
@kvuorine I see. Using DEVICE_COMPAT_VERSION won't work?

kvuorine commented:
I guess DEVICE_COMPAT_VERSION major version increment would prevent bootloops by preventing sysupgrade.

csharper2005 commented:
@kvuorine and one more idea. I guess that sysupgrade from small to new layout will work fine if you add additional partitions using `mtd-concat`. Here is an example https://github.com/openwrt/openwrt/blob/21f713d5abf86fc5639b41d7f4e7535a5538d63a/target/linux/ramips/dts/mt7621_tplink_ec330-g5u-v1.dts#L139. Of course, It won't escape from bootloop when we're sysupgrading from the new layout to the small one.

Znevna commented:
@csharper2005: I keep seeing `mtd-concat` recommended as a solution,  I don't know what use it has here, since the partitions that we want to merge are adjacent, there's nothing in between them, simply extending the first one will have the same effect, which sadly, without any helpers, is a bootlooper when sysupgrading from standard to `extended` builds (and backwards).
@Djfe found this: https://patchwork.ozlabs.org/project/openwrt/patch/20191112140214.59f2971f@kosmio.komorska/#2307317
It was never merged, but it could be a solution to use ubirsvol to safely upgrade from standard to extended partition.
Backwards I think only wipe and reinstall could be possible, so this might require a compat version increase and a big fat warning.

kvuorine commented:
If it  would be up to me, if sysupgrade functions as documentation suggests, I would just set ```DEVICE_COMPAT_VERSION := 2.0``` and something like ```DEVICE_COMPAT_MESSAGE := Sysupgrade is not possible due to partition layout change, proceed by installing factory image```, there are already documented method how to  do this and change the layout.

Djfe commented:
And Bob has easy way to recover
https://openwrt.org/toh/asus/rt-ax53u#back_to_asus_firmware_with_asus_repair_tool
Which is easier than the installation imo

It would be cool to figure out the ubi thing eventually but devs don't seem to have the time go look into it, so therefore not in this PR.
If you want to change this PR to change the original layout, then feel free.
the only issue I see is: you need windows and the repair tool to return to stock rom. currently you can copy the contents of firmware partition 2 to partition 1 and it works afaik.
though this was never mentioned in the wiki ^^

kvuorine commented:
I'm still doing testing, but installing factory image from OpenWRT with current nested layout seems to work using command ```mtd -r write factory.bin firmware```.

Znevna also has used tftp client when installing firmware instead of Asus repair tool.

I had to recover using bootloaders serial upgrade tools to install factory image when OpenWRT was damaged, possibly bootloader version 1.0.0.1 has nonfunctioning tftp tools.

kvuorine committed:
ramips: Asus RT-AX53U partition layout change

Change Asus RT-AX53U partition layout to maximize available
space.

Increase DEVICE_COMPAT_VERSION and add DEVICE_COMPAT_MESSAGE
to prevent unintentional damage.

Signed-off-by: Kimmo Vuorinen <kimmo.vuorinen@gmail.com>

kvuorine commented:
Install tests:

22.03.5 which has nested partitions backported ```mtd -r write openwrt-factory.bin firmware``` -> success.
late snapshot ```mtd -r write openwrt-factory.bin firmware``` -> success.
bootloader upgrader tftp download openwrt-factory.bin -> success.

```
# sysupgrade -n /tmp/openwrt-sysupgrade.bin 
Wed Jun  7 18:10:12 EEST 2023 upgrade: The device is supported, but this image is incompatible for sysupgrade based on the image version (1.1->2.0).
Wed Jun  7 18:10:12 EEST 2023 upgrade: Sysupgrade is not possible due to partition layout change, proceed by installing factory image
Image check failed.
```

Bob also tested sysupgrade -F and it failed ;)

Djfe commented:
I'm still fairly sure, that you should adjust `IMAGE_SIZE`, so people can build large images with lots of packages pre-installed :) (larger than 50MB)
The rest looks good to me

kvuorine commented:
I guess I'll stop aging nand on my device now, can't seem to find a way to install directly from OpenWRT prior partition nest commit.

Any views how large IMAGE_SIZE should be?

csharper2005 commented:
> Any views how large IMAGE_SIZE should be?

0x7420000 = 118912k

Additionally, you can take this into the account https://github.com/openwrt/openwrt/pull/11352#discussion_r1035032721

сс @DragonBluep

kvuorine commented:
@csharper2005 are you saying I should deduct physical eraseblocks reserved for ubi bad physical eraseblock handling and maybe some space for saving settings from ```IMAGE_SIZE```?

```
[    5.567092] UBI: auto-attach mtd7
[    5.570438] ubi0: attaching mtd7
[    8.445296] ubi0: scanning is finished
[    8.466731] ubi0: attached mtd7 (name "ubi", size 112 MiB)
[    8.472298] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    8.479148] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    8.485922] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    8.492881] ubi0: good PEBs: 897, bad PEBs: 0, corrupted PEBs: 0
[    8.498865] ubi0: user volume: 2, internal volumes: 1, max. volumes count: 128
[    8.506071] ubi0: max/mean erase counter: 2/0, WL threshold: 4096, image sequence number: 1686075034
[    8.515187] ubi0: available PEBs: 0, total reserved PEBs: 897, PEBs reserved for bad PEB handling: 19
[    8.524426] ubi0: background thread "ubi_bgt0d" started, PID 318
```

118912k - 2432k - (min rootfs_data)?

kvuorine commented:
I can't seem to stop wrecking the nand, from 23.03.3 install openwrt-factory.bin with:
```
dd if=/tmp/openwrt-factory.bin of=/dev/mtdblock4 bs=2048 count=2048
dd if=/tmp/openwrt-factory.bin of=/dev/mtdblock5 bs=2048 skip=2048
```
And power cycle the device.

I tried that earlier but somehow forgot to use block devices.

DragonBluep commented:
> > Any views how large IMAGE_SIZE should be?
> 
> 0x7420000 = 118912k
> 
> Additionally, you can take this into the account [#11352 (comment)](https://github.com/openwrt/openwrt/pull/11352#discussion_r1035032721)
> 
> сс @DragonBluep

At present, we do not have a good way to limit the size of nand images, and nand devices usually have enough space. To avoid complicating the problem, let's just keep the original values `118912k`.

DragonBluep commented:
> I can't seem to stop wrecking the nand, from 22.03.3 install openwrt-factory.bin with:
> 
> ```
> dd if=/tmp/openwrt-factory.bin of=/dev/mtdblock4 bs=2048 count=2048
> dd if=/tmp/openwrt-factory.bin of=/dev/mtdblock5 bs=2048 skip=2048
> ```
> 
> And power cycle the device.
> 
> I tried that earlier but somehow forgot to use block devices.

You need to erase mtdblock before writing it.

kvuorine commented:
> You need to erase mtdblock before writing it.

Well thats strange, I tested that procedure and it worked as is.

kvuorine commented:
I guess it would be good to ```mount -o remount,ro /overlay``` prior installing directly from OpenWRT.

I'm done testing, if this gets merged someday, should these ways to install be added into wiki device page?

I think at least if that tftp client as alternate of Asus repart tool should be added, if it is working but I can't confirm that.

remittor commented:
@kvuorine 
> 118912k - 2432k - (min rootfs_data)?

Not everything is clear here.
If you build an OpenWRT factory image larger than 60 MB, it will not fit into the already mounted `firmware` partition.
You need to ask the user to install a small factory image first. 
And only then can the user successfully install huge images, so that partition `firmware` will increase significantly.

You can also try installing the OpenWRT firmware via the WEB UI AsusWRT.
The new utility `mkasustrx` can create compatible `trx` images: https://github.com/openwrt-xiaomi/firmware-utils/commit/8eeb4e70aad5744acfc06bb5a0b1195af1cf07b4

kvuorine commented:
@remittor 
> The factory/initramfs image size should be limited to 50 MB. And sysupgrade image should be limited to 110 MB.

I see what you mean, maybe noting that factory image must be under 50MB in DEVICE_COMPAT_MESSAGE and install instructions from factory firmware would suffice?

matjon commented:
Hello,

I think that instead of having to flash factory image to enlarge UBI, it should be doable to erase UBI and format it again during sysupgrade. This would require running sysupgrade twice with patched OpenWRT, but should be better / quicker then flashing factory image. I mean something like that (rough, untested, based on OpenWRT 23.05.0):

```
diff --git a/target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh b/target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh
index 5d8305b788..d834c02831 100755
--- a/target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh
+++ b/target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh
@@ -47,6 +47,28 @@ platform_do_upgrade() {
                dd if=/dev/mtd5 bs=1024 count=52224 >> /tmp/backup_firmware.bin
                mtd -e firmware2 write /tmp/backup_firmware.bin firmware2
                ;;
+       asus,rt-ax53u)
+               ubidev=/sys/class/ubi/ubi0
+               total_eraseblocks="$(cat "$ubidev/total_eraseblocks")"
+               avail_eraseblocks="$(cat "$ubidev/avail_eraseblocks")"
+               reserved_for_bad="$(cat "$ubidev/reserved_for_bad")"
+               eraseblock_size="$(cat "$ubidev/eraseblock_size")"
+
+               volume_size=$(((total_eraseblocks + avail_eraseblocks + reserved_for_bad) * eraseblock_size))
+
+               # Thresholds are TBD
+               if [ $volume_size -lt $((50*1024*1024)) ] && [ $volume_size -gt $((1*1024*1024)) ]; then
+                       echo "Erasing UBI"
+                       # nand_detach_ubi() is defined in
+                       # package/base-files/files/lib/upgrade/nand.sh
+                       nand_detach_ubi ubi
+                       mtd erase ubi
+
+                       # nand_attach_ubi(), which is called by
+                       # nand_do_upgrade() will reformat the ubi, using new
+                       # (bigger) size
+               fi
+               ;;
        esac
 
        case "$board" in
```

kvuorine commented:
@matjon 
> I think that instead of having to flash factory image to enlarge UBI, it should be doable to erase UBI and format it again during sysupgrade. This would require running sysupgrade twice with patched OpenWRT, but should be better / quicker then flashing factory image.

That won't work, modified script is available after sysupgrade which has already broken the device.

Djfe commented:
what could be done:
make 23.05.1 and some master release increase compat version to 2.0

then make 23.05.2 have the required change to the ubi partition

That way compat version is increased before you require the larger compat version with the next release.
the note on the compat increase could inform users that they are required to sysupgrade to 23.05.01 (or some snapshot) first, before being able to update to 23.05.2

Djfe commented:
or is the issue that the sysupgrade is calculating partition size based on the old dts file?
in that case we might have to introduce a custom build intermediate sysupgrade build for people to install(?)

remittor commented:
> That way compat version is increased before you require the larger compat version with the next release.

How to block an attempt to write to partition `Kernel` of a huge image (more than 50 MB)?

remittor commented:
**Offtopic**
OpenWRT maintainers inadvertently limited the maximum image size to 16 MiB for routers TUF-AX4200 and TUF-AX6000:
https://github.com/openwrt/openwrt/pull/13422#issuecomment-1780773692

And 50 MB is not enough for you...

kvuorine commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -342,7 +342,10 @@ define Device/asus_rt-ax53u
+  DEVICE_COMPAT_VERSION := 2.0
+  DEVICE_COMPAT_MESSAGE := Sysupgrade is not possible due to partition \
+	layout change, proceed by installing factory image
+  IMAGE_SIZE := 118912k
```
I haven't made any decisions about that with my question following the issue you brought up.

If you look at earlier discussions in the thread you'll see that it was not a part of the original pull request but changed after a discussion about that.

I don't see the need for such a large read only root as updating software is more sane when it already resides in writable partition but I also don't see the need to restrict users if someone wants to do that if it doesn't cause significant issues.

User could install such a huge factory or sysupgrade image after initially installing smaller image with a new partition layout which fits into the stock partition.

kvuorine commented:
> or is the issue that the sysupgrade is calculating partition size based on the old dts file? in that case we might have to introduce a custom build intermediate sysupgrade build for people to install(?)

Probably the issue is with sysupgrade, it doesn't have the data for new partitions.
